src/awkward1/operations/convert.py

```diff
@@ -1139,6 +1139,180 @@ def getfunction(layout, depth):
+
+            if (awkward1.operations.describe.parameters(layout).get("__array__") == "bytestring"):
+                arr = [awkward1.behaviors.string.ByteBehavior(
+                    layout[i]).__bytes__() for i in range(len(layout))]
+                if mask is not None:
+                    return pyarrow.array(arr, mask=numpy.array(mask, dtype=numpy.bool))
+                else:
+                    return pyarrow.array(arr) 
+
+            if (awkward1.operations.describe.parameters(layout).get("__array__") == "string"):
+                arr = [awkward1.behaviors.string.CharBehavior(
+                    layout[i]).__str__() for i in range(len(layout))]
+                if mask is not None:
+                    return pyarrow.array(arr, mask=numpy.array(mask, dtype=numpy.bool))
+                else:
+                    return pyarrow.array(arr) 
+            
+            if mask is not None:
+                indices = []
+                for i in range(len(mask)):
+                    if mask[i] == 1:
+                        indices.append(None)
+                    else:
+                        indices.append(i)
+                return pyarrow.LargeListArray.from_arrays(recurse(layout.offsets), 
+                                                            recurse(layout.content)).take( pyarrow.array(indices))
+            
+            return pyarrow.LargeListArray.from_arrays(recurse(layout.offsets), recurse(layout.content))
+        
+        elif isinstance(layout, (awkward1.layout.RegularArray)):
+            return recurse(layout.broadcast_tooffsets64(layout.compact_offsets64()))
+
+        elif isinstance(layout, (awkward1.layout.ListArray32, 
+                                awkward1.layout.ListArrayU32, 
+                                awkward1.layout.ListArray64)):
+            if mask is not None:
+                return recurse(layout.broadcast_tooffsets64(layout.compact_offsets64()), mask)
+            else:
+                return recurse(layout.broadcast_tooffsets64(layout.compact_offsets64()))
+
+        elif isinstance(layout, (awkward1.layout.RecordArray)):
+            values = [recurse(x) for x in layout.contents]
+            # Arrow rejects the values list if every child array is not of same length
+            min_list_len = min(map(len, values))
+
+            if mask is not None:
+                for i in range(len(values)):
+                    values[i] = values[i].slice(0, len(mask)) 
+                indices = []
+                for i in range(len(mask)):
+                    if mask[i] == 1:
+                        indices.append(None)
+                    else:
+                        indices.append(i)   
+                return pyarrow.StructArray.from_arrays(values, layout.keys()).take(pyarrow.array(indices))
+            else:
+                for i in range(len(values)):
+                    values[i] = values[i].slice(0, min_list_len)
```

Comment:
I needed to slice the sub-arrays since Array rejects unequal child length arrays.